File: Factory Method/Exemplos/Ex3_Nomes/Ex5.py
"""
PADRÃO: FACTORY METHOD (MÉTODO FÁBRICA)

Este exemplo demonstra como esconder a complexidade da criação de objetos 
(como o parsing de strings) atrás de um método fábrica.
"""

from abc import ABC

# 1. PRODUTO ABSTRATO
class Nome(ABC):
    def __init__(self):
        self.nome = ""
        self.sobrenome = ""

    def __str__(self):
        return f"Nome: {self.nome} | Sobrenome: {self.sobrenome}"

# 2. PRODUTOS CONCRETOS
class NomeSimples(Nome):
    def __init__(self, s):
        super().__init__()
        self.nome = s

class NomeSobrenome(Nome):
    def __init__(self, s):
        super().__init__()
        if "," in s:
            parts = s.split(",")
            self.sobrenome = parts[0].strip()
            self.nome = parts[1].strip()

# 3. CRIADOR (Fábrica)
class FabricaDeNomes:
    # Factory Method
    def get_nome(self, s) -> Nome:
        if "," in s:
            return NomeSobrenome(s)
        else:
            return NomeSimples(s)

# 4. CLIENTE
if __name__ == "__main__":
    fabrica = FabricaDeNomes()

    n1 = fabrica.get_nome("Silvio Santos")
    n2 = fabrica.get_nome("Lula, Luiz Inácio")

    print(n1)
    print(n2)

File: Factory Method/Exemplos/Ex3_Nomes/Ex5.js
/**
 * PADRÃO: FACTORY METHOD (MÉTODO FÁBRICA)
 * 
 * O Factory Method é útil para encapsular a lógica de criação de objetos 
 * quando essa criação depende do formato dos dados de entrada.
 */

// 1. PRODUTO ABSTRATO (Base)
class Nome {
    constructor() {
        this.nome = "";
        this.sobrenome = "";
    }

    toString() {
        return `Nome: ${this.nome} | Sobrenome: ${this.sobrenome}`;
    }
}

// 2. PRODUTOS CONCRETOS
class NomeSimples extends Nome {
    constructor(s) {
        super();
        this.nome = s;
    }
}

class NomeSobrenome extends Nome {
    constructor(s) {
        super();
        const parts = s.split(",");
        if (parts.length > 1) {
            this.sobrenome = parts[0].trim();
            this.nome = parts[1].trim();
        }
    }
}

// 3. CRIADOR (Fábrica)
class FabricaDeNomes {
    // Factory Method
    getNome(s) {
        if (s.includes(",")) {
            return new NomeSobrenome(s);
        } else {
            return new NomeSimples(s);
        }
    }
}

// 4. CLIENTE
const fabrica = new FabricaDeNomes();

const n1 = fabrica.getNome("Silvio Santos");
const n2 = fabrica.getNome("Lula, Luiz Inácio");

console.log(n1.toString());
console.log(n2.toString());
